### PR TITLE
Fixed removing multiple listeners within event callbacks #5827

### LINKED ISSUE #5827: Event.removeEventListener removes listeners incorrectly
When removing *multiple* listeners from an instance of`Event`,  **while the event is being raised**, the listeners are not removed correctly. 

See:

```js
var event = new Cesium.Event();

// add a bunch of listeners, but the even ones should 
// remove themselves the first time they are called. 
['A','B','C','D','E','F','G'].forEach((value, index) => {
	var remove = e.addEventListener(() => { 
                var isEven = index % 2 === 0;
                console.log(value, index, isEven ? '--removing this one!' : ''); 
                if (isEven) remove(); 
        });
});

// raise the event!
e.raiseEvent(); 
```

Outputs the following (as expected):
```
A 0 --removing this one!
B 1
C 2 --removing this one!
D 3
E 4 --removing this one!
F 5
G 6 --removing this one!
```

But when calling `e.raiseEvent()` again, we see:
```
B 1
F 5
```

instead of the expected:

```
B 1
D 3
F 5
```

### PR DESCRIPTION
**Pull request by speigg**
Fixed a bug that occurs when removing multiple event listeners within an event callback (see #5827). Added a test case to EventSpec.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
@speigg, thanks for the pull request! Maintainers, we have a signed CLA from @speigg, so you can review this at any time.

I noticed that [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md) has not been updated. If this change updates the public API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and comment on this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).



__I am a bot who helps you make Cesium awesome!__ Thanks again.

**Comment by ggetz** [comment]
Thanks for the PR @speigg !

If you add `Fixes #5827` to the description, the `Fixes` keyword will automatically close that issue when we merge this PR.

File: CHANGES.md [hunk 1]
```diff
@@ -1,5 +1,8 @@
 Change Log
 ==========
+
+* Fixed removing multiple event listeners within event callbacks. [#5827](https://github.com/AnalyticalGraphicsInc/cesium/issues/5827)
```
Comment (ggetz) [inline CHANGES.md]: Add this bullet under the next release, under ` ### 1.38 - 2017-10-02`

File: Source/Core/Event.js [hunk 2]
```diff
@@ -120,6 +120,10 @@ define([
         return false;
     };
 
+    function compareNumber(a,b) {
```
Comment (ggetz) [inline Source/Core/Event.js]: Since this is so short, it can probably be inline.


```
toRemove.sort(function (a, b) {
    ...
});
```

File: Specs/Core/EventSpec.js [hunk 3]
```diff
@@ -67,6 +67,29 @@ defineSuite([
...
     });
 
+    it('can remove multiple listeners within a callback', function() {
+        var listeners = [];
+
+        function addListener(index) {
+            var remove;
+            var listener = function() {
+                if (index % 2 === 0) {
+                    remove();
```
Comment (ggetz) [inline Specs/Core/EventSpec.js]: This test is a little hard to follow. Can you try structuring it a little more linearly like the other tests?

File: Source/Core/Event.js (hunk 2 above)
Comment (speigg) [reply Source/Core/Event.js]: Wouldn't that add an unnecessary callback allocation every time an event is raised?

File: Specs/Core/EventSpec.js (hunk 3 above)
Comment (speigg) [reply Specs/Core/EventSpec.js]: Sure

**Comment by speigg** [comment]
@ggetz I've rewritten the test case to make it simpler, as you suggested

File: Source/Core/Event.js (hunk 2 above)
Comment (ggetz) [reply Source/Core/Event.js]: That's true, fair enough!

**Comment by ggetz** [comment]
Thanks @speigg !

@pjcozzi This looks good to me, did you want any additional reviewers?

**Comment by pjcozzi** [comment]
Thanks @speigg @ggetz!
